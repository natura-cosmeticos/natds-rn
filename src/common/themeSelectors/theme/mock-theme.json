{
  "avatarSizes": {
    "tiny": {
      "size": 24,
      "fontSize": ".625rem"
    },
    "small": {
      "size": 32,
      "fontSize": ".875rem"
    },
    "standard": {
      "size": 40,
      "fontSize": "1rem"
    },
    "large": {
      "size": 48,
      "fontSize": "1rem"
    },
    "huge": {
      "size": 80,
      "fontSize": "1.5rem"
    }
  },
  "buttonSizes": {
    "small": {
      "paddingTop": 8,
      "paddingRight": 8,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "minHeight": 32
    },
    "medium": {
      "paddingTop": 12,
      "paddingRight": 16,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "minHeight": 40
    },
    "large": {
      "paddingTop": 16,
      "paddingRight": 16,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "minHeight": 48
    }
  },
  "fontSize": {
    "h1": {
      "fontSize": "6rem",
      "fontWeight": 300
    },
    "h2": {
      "fontSize": "3.75rem",
      "fontWeight": 300
    },
    "h3": {
      "fontSize": "3rem",
      "fontWeight": 400
    },
    "h4": {
      "fontSize": "2.125rem",
      "fontWeight": 400
    },
    "h5": {
      "fontSize": "1.5rem",
      "fontWeight": 400
    },
    "h6": {
      "fontSize": "1.25rem",
      "fontWeight": 600
    },
    "subtitle1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "subtitle2": {
      "fontSize": "0.875rem",
      "fontWeight": 500
    },
    "body1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "body2": {
      "fontSize": "0.875rem",
      "fontWeight": 400
    },
    "button": {
      "fontSize": "0.875rem",
      "fontWeight": 600
    },
    "caption": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    },
    "overline": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    }
  },
  "iconSizes": {
    "micro": 16,
    "tiny": 24,
    "small": 32,
    "standard": 48,
    "medium": 64,
    "large": 72,
    "largex": 96,
    "largexx": 128,
    "huge": 144,
    "hugex": 192,
    "hugexx": 256
  },
  "opacity": {
    "12": 0.12,
    "16": 0.16,
    "24": 0.24,
    "32": 0.32,
    "48": 0.48,
    "56": 0.56,
    "64": 0.64,
    "80": 0.8,
    "00": 0,
    "04": 0.04,
    "08": 0.08
  },
  "overlay": {
    "solidBlack": 0.56
  },
  "radius": {
    "none": 0,
    "circle": "50%",
    "small": 2,
    "medium": 4,
    "large": 8
  },
  "sizes": {
    "none": 0,
    "micro": 4,
    "tiny": 8,
    "small": 16,
    "standard": 24,
    "semi": 32,
    "semix": 40,
    "medium": 48,
    "mediumx": 56,
    "large": 64,
    "largex": 72,
    "largexx": 80,
    "largexxx": 88,
    "huge": 96,
    "hugex": 128,
    "hugexx": 144,
    "hugexxx": 192,
    "veryhuge": 256
  },
  "spacing": {
    "spacing": 8,
    "spacingNone": 0,
    "spacingMicro": 4,
    "spacingTiny": 8,
    "spacingSmall": 16,
    "spacingStandard": 24,
    "spacingSemi": 32,
    "spacingLarge": 48,
    "spacingHuge": 64
  },
  "colorTokens": {
    "colorPrimary": "#F4AB34",
    "colorOnPrimary": "#333333",
    "colorPrimaryLight": "#FEFDE8",
    "colorOnPrimaryLight": "#333333",
    "colorPrimaryDark": "#EF8426",
    "colorOnPrimaryDark": "#333333",
    "colorSecondary": "#FF6B0B",
    "colorOnSecondary": "#000000",
    "colorSecondaryLight": "#FFF8E1",
    "colorOnSecondaryLight": "#000000",
    "colorSecondaryDark": "#FF5808",
    "colorOnSecondaryDark": "#000000",
    "colorBackground": "#FAFAFA",
    "colorOnBackground": "#333333",
    "colorSurface": "#FFFFFF",
    "colorOnSurface": "#333333",
    "colorHighlight": "#000000",
    "colorHighEmphasis": "#333333",
    "colorMediumEmphasis": "#777777",
    "colorLowEmphasis": "#BBBBBB",
    "colorLink": "#227BBD",
    "colorOnLink": "#FFFFFF",
    "colorSuccess": "#569A32",
    "colorOnSuccess": "#FFFFFF",
    "colorWarning": "#FCC433",
    "colorOnWarning": "#333333",
    "colorAlert": "#E74627",
    "colorOnAlert": "#FFFFFF"
  },
  "elevation": {
    "1": {
      "elevation": 1,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 1, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 1
    },
    "2": {
      "elevation": 2,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 2, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 2
    },
    "3": {
      "elevation": 3,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 3, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 4
    },
    "4": {
      "elevation": 4,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 4, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 5
    },
    "5": {
      "elevation": 5,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 6, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "6": {
      "elevation": 6,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 8, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "7": {
      "elevation": 7,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 9, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 12
    },
    "8": {
      "elevation": 8,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 12, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 17
    },
    "9": {
      "elevation": 9,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 16, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 24
    },
    "10": {
      "elevation": 10,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 24, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 38
    }
  }
}
