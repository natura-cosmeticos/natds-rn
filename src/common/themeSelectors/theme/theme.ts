import { tokens as styleTokens } from '@naturacosmeticos/natds-styles';
import { IColorThemeTokens } from '@naturacosmeticos/natds-styles/dist/tokens/colors/themeBound/IColorThemeTokens';
import { IElevation } from '@naturacosmeticos/natds-styles/dist/tokens/elevation/IElevation';

type Tokens = typeof styleTokens;
type TokensWithoutColor = Omit<Tokens, 'colors'>
type TokensWithoutColorTokens = Omit<TokensWithoutColor, 'colorTokens'>
type TokensWithoutElevation = Omit<TokensWithoutColorTokens, 'elevation'>

interface ElevationProps {
  elevation: number
  shadowColor: string
  shadowOffset: { height: number, width: number }
  shadowOpacity: number
  shadowRadius: number
}

type Elevation = Record<keyof IElevation, ElevationProps>


export type Theme = TokensWithoutElevation
  & {
    colorTokens: IColorThemeTokens,
    elevation: Elevation | {}
  }

type ThemeMode = 'dark' | 'light';

const parseElevation = (elevation: string) => {
  const regEx = /(\d+(\.\d+)?)/g;
  const matches = elevation.match(regEx);

  if (matches) {
    const [width, height, blurRadius, , r, g, b, alpha] = matches;

    const toUnit = (prop: string) => parseInt(prop, 0);

    return {
      elevation: 2,
      shadowColor: `rgb(${r}, ${g}, ${b})`,
      shadowOffset: { height: toUnit(height), width: toUnit(width) },
      shadowOpacity: parseFloat(alpha),
      shadowRadius: toUnit(blurRadius),
    };
  }

  return {};
};

const buidElevation = (elevation: IElevation) => {
  const nativeElevation = {};

  Object
    .keys(elevation)
    .forEach((key) => { nativeElevation[key] = parseElevation(elevation[key]); });

  return nativeElevation;
};

export const buildTheme = (tokens: Tokens, brand: string, mode: ThemeMode) => {
  const colorThemeName = `${brand}${mode === 'light' ? 'Light' : 'Dark'}ColorTokens`;
  const localTokens = { ...tokens };

  delete localTokens.colors;

  const theme: Theme = {
    ...localTokens,
    colorTokens: localTokens.colorTokens[colorThemeName],
    elevation: buidElevation(localTokens.elevation),
  };

  return theme;
};

interface ThemeSelector {
  (theme: Theme): string | undefined
}

export const getTheme = (selector: ThemeSelector) => (
  { theme }: { theme: Theme },
) => selector(theme);
