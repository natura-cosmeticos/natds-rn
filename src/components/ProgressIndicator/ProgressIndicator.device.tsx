import { storiesOf } from '@storybook/react-native'

import { all, interactive } from './ProgressIndicator.stories'

storiesOf('ProgressIndicator', module).add('all', all).add('interactive', interactive)
