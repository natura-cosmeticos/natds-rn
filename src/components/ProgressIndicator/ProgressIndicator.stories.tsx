import React from 'react';
import { View } from 'react-native';
import { select } from '@storybook/addon-knobs';

import { ContainerWithTheme } from '../../common/HelperComponents/ThemeHelper.styles';
import {
  ProgressIndicator,
  ProgressIndicatorComponent,
  ProgressIndicatorVariants,
  ProgressIndicatorTypes,
  ProgressIndicatorColors,
} from './ProgressIndicator';

export default {
  component: ProgressIndicatorComponent,
  parameters: {
    componentSubtitle: 'Progress indicators express an unspecified wait time or display the length of a process.',
  },
  title: 'Components|ProgressIndicator',
};

const variantTypes = {
  variant: 'circular',
};
const types = {
  indeterminate: 'indeterminate',
};
const colorTypes = {
  primary: 'primary',
  secondary: 'secondary',
};

export const circular = () => (
  <View style={{ maxWidth: 600 }}>
    <ContainerWithTheme>
      <View style={{ margin: 15 }}>
        <ProgressIndicator variant="circular" type="indeterminate" color="primary" />
      </View>
      <View style={{ margin: 15 }}>
        <ProgressIndicator variant="circular" type="indeterminate" color="secondary" />
      </View>
    </ContainerWithTheme >
  </View>
);

export const interactive = () => (
  <View style={{ maxWidth: 600 }}>
    <ContainerWithTheme>
      <View style={{ margin: 15 }}>
        <ProgressIndicator
          variant={select('Variant', variantTypes, 'circular') as ProgressIndicatorVariants}
          type={select('Type', types, 'indeterminate') as ProgressIndicatorTypes}
          color={select('Color', colorTypes, 'primary') as ProgressIndicatorColors}
        />
      </View>
    </ContainerWithTheme >
  </View>
);
