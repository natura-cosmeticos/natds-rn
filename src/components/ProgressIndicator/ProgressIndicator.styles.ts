/* eslint-disable max-len */
import { StyleSheet } from 'react-native'
import styled from 'styled-components/native'
import { Theme, getColorPrimary, getColorSurface } from '../../common/themeSelectors'

export const View = styled.View<{size: number; showLayer: boolean; theme: Theme}>(
  ({ size, showLayer, theme }) => ({
    alignItems: 'center',
    backgroundColor: showLayer ? getColorSurface(theme) : 'transparent',
    borderRadius: (size + 4) / 2,
    display: 'flex',
    height: size + 4,
    justifyContent: 'center',
    width: size + 4
  })
)

export const Layer = styled.View<{
  size: number;
}>(({ size }) => ({
  height: size,
  width: size
}))

export const Line = styled.View<{ theme: Theme; size: number }>(
  ({ theme, size }) => ({
    borderColor: getColorPrimary(theme),
    borderRadius: size / 2,
    borderWidth: size / 10,
    height: size,
    width: size
  })
)

export const Container = styled.View<{
  size: number;
}>(({ size }) => ({
  height: size / 2,
  overflow: 'hidden',
  width: size
}))

export const Loop = styled.View(() => ({
  ...StyleSheet.absoluteFillObject,
  alignItems: 'center',
  justifyContent: 'center'
}))
