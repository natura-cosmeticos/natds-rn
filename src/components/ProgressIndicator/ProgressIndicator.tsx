import React from 'react';
import { tokens } from '@naturacosmeticos/natds-styles';

import { CircularProgress } from './CircularProgress/CircularProgress';

export type ProgressIndicatorVariants = 'circular';
export type ProgressIndicatorColors = 'primary' | 'secondary';
export type ProgressIndicatorTypes = 'indeterminate';

export interface ProgressIndicatorProps {
  /**
  * Variant of the progress indicator: `circular`
  */
  variant: ProgressIndicatorVariants;
  /**
  * Color of the progress indicator: `primary` | `secondary`
  */
  color?: ProgressIndicatorColors;
  /**
  *  Type of the progress indicator: `indeterminate`
  */
  type?: ProgressIndicatorTypes,
  /**
   * Size to be used on circular progress
   */
  size?: number;
}

export const ProgressIndicatorComponent = ({
  variant,
  color = 'primary',
  type = 'indeterminate',
  size = tokens.sizes.standard,
}: ProgressIndicatorProps) => {
  if (type !== 'indeterminate' || variant !== 'circular') {
    throw new Error('Not implemented yet');
  }

  return (
    <CircularProgress
      size={size}
      color={color}
    />
  );
};

export const ProgressIndicator = React.memo(ProgressIndicatorComponent);
