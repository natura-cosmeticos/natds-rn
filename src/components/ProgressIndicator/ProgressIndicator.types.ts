export type ProgressIndicatorSizes = 'standard' | 'semi' | 'medium' | 'large';

export interface ProgressIndicatorProps {
  /**
   * Size to be used in the progress indicator
   */
  size?: ProgressIndicatorSizes;
  /**
   * Shows a colored layer behind the progress indicator
   */
  showLayer?: boolean;
}
