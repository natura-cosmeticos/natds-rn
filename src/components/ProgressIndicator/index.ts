export { ProgressIndicator, ProgressIndicatorVariants, ProgressIndicatorColors, ProgressIndicatorTypes } from './ProgressIndicator';
