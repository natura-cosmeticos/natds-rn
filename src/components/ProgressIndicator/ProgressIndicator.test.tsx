import React from 'react';

import { ThemeProvider } from 'styled-components/native';
import renderer from 'react-test-renderer';

import theme from '../../common/themeSelectors/theme/mock-theme.json';
import { ProgressIndicator, ProgressIndicatorProps } from './ProgressIndicator';

jest.mock('./CircularProgress/CircularProgress', () => ({
  CircularProgress: 'CircularProgress',
}));

const defaultProps = {
  color: 'primary',
  type: 'indeterminate',
  variant: 'circular',
} as Omit<ProgressIndicatorProps, 'theme'>;

const renderProgressIndicator = (props?: Omit<ProgressIndicatorProps, 'theme'>) => (renderer.create(
  <ThemeProvider theme={theme}>
    <ProgressIndicator
      {...defaultProps}
      {...props}
    />
  </ThemeProvider>,
));

describe('ProgressIndicator component', () => {
  it('Should throw error when trying to render with non supported type or variant', () => {
    const renderProgressIndicatorWithError = () => renderProgressIndicator({
      ...defaultProps,
      type: 'unknown-type' as any,
      variant: 'unknown-variant' as any,
    });

    const error = new Error('Not implemented yet');

    expect(renderProgressIndicatorWithError).toThrow(error);
  });

  describe('circular', () => {
    it('Should render circular progress correctly', () => {
      const circularProgress = renderProgressIndicator().toJSON();

      expect(circularProgress).toMatchSnapshot();
    });
  });
});
