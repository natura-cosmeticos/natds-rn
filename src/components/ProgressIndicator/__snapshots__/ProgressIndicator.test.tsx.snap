// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressIndicator component circular Should render circular progress correctly 1`] = `
<CircularProgress
  color="primary"
  size={24}
/>
`;
